src/core/Akka/Actor/FSM.cs

```diff
@@ -1017,73 +1097,82 @@ private static StateFunction OrElse(StateFunction original, StateFunction fallba
 
         #region Actor methods
 
-        /// <summary>
-        /// Main actor receive method
-        /// </summary>
-        /// <param name="message">TBD</param>
-        /// <returns>TBD</returns>
+        /// <inheritdoc/>
         protected override bool Receive(object message)
         {
-            var match = PatternMatch.Match(message)
-                .With<TimeoutMarker>(marker =>
-                {
-                    if (_generation == marker.Generation)
-                    {
-                        ProcessMsg(new StateTimeout(), "state timeout");
-                    }
-                })
-                .With<Timer>(t =>
-                {
-                    if (_timers.ContainsKey(t.Name) && _timers[t.Name].Generation == t.Generation)
-                    {
-                        if (_timeoutFuture != null)
-                        {
-                            _timeoutFuture.Cancel(false);
-                            _timeoutFuture = null;
-                        }
-                        _generation++;
-                        if (!t.Repeat)
-                        {
-                            _timers.Remove(t.Name);
-                        }
-                        ProcessMsg(t.Message,t);
-                    }
-                })
-                .With<SubscribeTransitionCallBack>(cb =>
-                {
-                    Context.Watch(cb.ActorRef);
-                    Listeners.Add(cb.ActorRef);
-                    //send the current state back as a reference point
-                    cb.ActorRef.Tell(new CurrentState<TState>(Self, _currentState.StateName));
-                })
-                .With<Listen>(l =>
-                {
-                    Context.Watch(l.Listener);
-                    Listeners.Add(l.Listener);
-                    l.Listener.Tell(new CurrentState<TState>(Self, _currentState.StateName));
-                })
-                .With<UnsubscribeTransitionCallBack>(ucb =>
-                {
-                    Context.Unwatch(ucb.ActorRef);
-                    Listeners.Remove(ucb.ActorRef);
-                })
-                .With<Deafen>(d =>
+            var timeoutMarker = message as TimeoutMarker;
+            if (timeoutMarker != null)
+            {
+                if (_generation == timeoutMarker.Generation)
                 {
-                    Context.Unwatch(d.Listener);
-                    Listeners.Remove(d.Listener);
-                })
-                .With<InternalActivateFsmLogging>(_=> { DebugEvent = true; })
-                .Default(msg =>
+                    ProcessMsg(StateTimeout.Instance, "state timeout");
```

Comment:
Way easier to read (the main FSM handling routine) and will have fewer allocations also. Nice job.